compiler/next/include/chpl/uast/On.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_UAST_ON_H
+#define CHPL_UAST_ON_H
+
+#include "chpl/queries/Location.h"
+#include "chpl/uast/BlockStyle.h"
+#include "chpl/uast/Expression.h"
+#include "chpl/uast/SimpleBlockLike.h"
+
+namespace chpl {
+namespace uast {
+
+
+/**
+  This class represents an on statement. For example:
+
+  \rst
+  .. code-block:: chapel
+
+      // Example 1:
+      var x = 0;
+      on Locales[1] do writeln(x);
+
+  \endrst
+
+ */
+class On final : public SimpleBlockLike {
+ private:
+  On(ASTList children, BlockStyle blockStyle, int bodyChildNum,
+     int numBodyStmts)
+    : SimpleBlockLike(asttags::On, std::move(children), blockStyle,
+                      bodyChildNum,
+                      numBodyStmts) {
+    assert(isExpressionASTList(children_));
+  }
+
+  bool contentsMatchInner(const ASTNode* other) const override {
+    return simpleBlockLikeContentsMatchInner(other);
+  }
+
+  void markUniqueStringsInner(Context* context) const override {
+    simpleBlockLikeMarkUniqueStringsInner(context);
+  }
+
+  static const int8_t destChildNum_ = 0;
+
+ public:
+
+  /**
+    Create and return an on statement.
+  */
+  static owned<On> build(Builder* builder, Location loc,
+                         owned<Expression> destination,
+                         BlockStyle blockStyle,
+                         ASTList stmts);
+
+  /**
+    Returns the destination of this on statement.
+  */
+  const Expression* destination() const {
```

Comment:
Perhaps this method should be called `locale()` ?

Reply:
In talking to @vasslitvinov I think we prefer `destination()` because it talks more about the purpose of the expression, whereas when I read `locale()` I tend to think of the expression type.

Reply:
Still sitting on this method name. @mppf @vasslitvinov out of `destination`, `target`, `locale` which would you prefer? Care to take a (by post) vote?

Reply:
I don't have a strong opinion here. I think I brought up alternative naming to `destination` because that term is relatively ambiguous (is it about where the uAST node will end up after parsing, say?). I have the same feeling about `target` but slightly less so. If we did not think it was too long we could combine the ideas - `targetLocale` or `destinationLocale`. But I will be OK with any of these.